src/array/dictionary.rs

```diff
@@ -0,0 +1,119 @@
+use std::sync::Arc;
+
+use crate::{
+    buffer::{types::NativeType, Bitmap},
+    datatypes::DataType,
+};
+
+use super::{ffi::ToFFI, new_empty_array, primitive::PrimitiveArray, Array};
+
+pub unsafe trait DictionaryKey: NativeType {
+    const DATA_TYPE: DataType;
+}
+
+unsafe impl DictionaryKey for i8 {
+    const DATA_TYPE: DataType = DataType::Int8;
+}
+unsafe impl DictionaryKey for i16 {
+    const DATA_TYPE: DataType = DataType::Int16;
+}
+unsafe impl DictionaryKey for i32 {
+    const DATA_TYPE: DataType = DataType::Int32;
+}
+unsafe impl DictionaryKey for i64 {
+    const DATA_TYPE: DataType = DataType::Int64;
+}
+unsafe impl DictionaryKey for u8 {
+    const DATA_TYPE: DataType = DataType::UInt8;
+}
+unsafe impl DictionaryKey for u16 {
+    const DATA_TYPE: DataType = DataType::UInt16;
+}
+unsafe impl DictionaryKey for u32 {
+    const DATA_TYPE: DataType = DataType::UInt32;
+}
+unsafe impl DictionaryKey for u64 {
+    const DATA_TYPE: DataType = DataType::UInt64;
+}
+
+#[derive(Debug, Clone)]
+pub struct DictionaryArray<K: DictionaryKey> {
+    data_type: DataType,
+    keys: PrimitiveArray<K>,
```

Comment:
this is really cool (that dictionaries have `keys`, `value` directly as fields rather than knowing `child[0]` is keys and `child[1]` is values